### ISSUE TITLE
Mat_VarFree of a duplicated variable causes a double free segfault

### ISSUE THREAD

**Issue by S-Dafarra** [tbeu/matio#158]
If I run the following snippet
```c++
#include <vector>
#include <matio.h>

int main()
{

    std::vector<double> vec = {1, 2, 3, 5};
    std::vector<size_t> dimensions = {vec.size(), 1};
    matvar_t* matioVar = Mat_VarCreate("test", matio_classes::MAT_C_DOUBLE, matio_types::MAT_T_DOUBLE, dimensions.size(), dimensions.data(), vec.data(), 0);

    matvar_t * shallowCopy = Mat_VarDuplicate(matioVar, 0);

    Mat_VarFree(shallowCopy);
    Mat_VarFree(matioVar);

    return EXIT_SUCCESS;
}
```
it segfaults with the error
```
free(): double free detected in tcache 2
```
If I set the ``data`` field of ``shallowCopy`` to ``nullptr`` the problem is avoided, i.e.
```c++
#include <vector>
#include <matio.h>

int main()
{

    std::vector<double> vec = {1, 2, 3, 5};
    std::vector<size_t> dimensions = {vec.size(), 1};
    matvar_t* matioVar = Mat_VarCreate("test", matio_classes::MAT_C_DOUBLE, matio_types::MAT_T_DOUBLE, dimensions.size(), dimensions.data(), vec.data(), 0);

    matvar_t * shallowCopy = Mat_VarDuplicate(matioVar, 0);

    shallowCopy->data = nullptr;
    Mat_VarFree(shallowCopy);
    Mat_VarFree(matioVar);

    return EXIT_SUCCESS;
}
```
runs with no problem. Is this an expected behavior?

**Comment by tbeu** [comment]
Thanks for reporting. I guess matio needs to keep track of the reference count in Mat_VarDuplicate/Mat_VarFree. Currently this not the case.

**Comment by S-Dafarra** [comment]
> Thanks for reporting. I guess matio needs to keep track of the reference count in Mat_VarDuplicate/Mat_VarFree. Currently this not the case.

Ah, I see. So calling ``Mat_VarDuplicate`` with 0 "shares" the memory. Right now the first variable that gets freed deallocates ``data``. I was somehow expecting the original variable to own ``data``, but that was my interpretation. 

I saw later this https://github.com/tbeu/matio/blob/260a830c4f60e70df0b3ba22cdfbaad9431c0faf/src/mat.c#L1376-L1378

The thing I did not understand is that ``Mat_VarFree`` needs to be called also in the shallow copy because the other content of the struct needs to be deallocated.
